api/src/context.rs

```diff
@@ -181,104 +178,39 @@ impl Context {
+                    first_version,
+                );
             }
-        }
-
-        // If timestamp is 0, it's the genesis transaction, and we can stop now
-        if timestamp == 0 {
-            return Ok(BlockInfo {
-                block_height: 0,
-                start_version: start,
-                end_version: end,
-                block_hash: block_hash.into(),
-                block_timestamp: timestamp,
-                num_transactions: end.saturating_sub(start).saturating_add(1) as u16,
-            });
-        }
-
-        // Retrieve block height from the transaction outputs
-        let height_id = ident_str!("height");
-        let block_metadata_type = move_deps::move_core_types::language_storage::StructTag {
-            address: CORE_CODE_ADDRESS,
-            module: ident_str!("block").into(),
-            name: ident_str!("BlockMetadata").into(),
-            type_params: vec![],
         };
 
-        let resolver = self.move_resolver()?;
-        let converter = resolver.as_converter(self.db.clone());
-        let txn = self.get_transaction_by_version(start, ledger_version)?;
-
-        // Parse the resources and find the block metadata resource update
-        let maybe_block_height = txn.changes.iter().find_map(|(key, op)| {
-            if let StateKey::AccessPath(path) = key {
-                if let Path::Resource(typ) = path.get_path() {
-                    // If it's block metadata, we can convert it to get the block height
-                    // And it must be the root address
-                    if path.address == CORE_CODE_ADDRESS && typ == block_metadata_type {
-                        if let WriteOp::Value(value) = op {
-                            if let Ok(mut resource) = converter.try_into_resource(&typ, value) {
-                                if let Some(value) = resource.data.0.remove(&height_id.into()) {
-                                    if let Ok(height) = serde_json::from_value::<U64>(value) {
-                                        return Some(height.0);
-                                    }
-                                }
-                            }
-                        }
-                    }
-                }
-            }
-
-            None
-        });
-
-        // This should always work unless there's something unexpected in the block format
-        if let Some(block_height) = maybe_block_height {
-            Ok(BlockInfo {
-                block_height,
-                start_version: start,
-                end_version: end,
-                block_hash: block_hash.into(),
-                block_timestamp: timestamp,
-                num_transactions: end.saturating_sub(start).saturating_add(1) as u16,
-            })
-        } else {
-            Err(anyhow!(
-                "Unable to find block height in metadata transaction {}:{}",
-                start,
-                end
-            ))
-        }
+        Ok(BlockInfo {
```

Comment:
We can probably define a "clean" BlockInfo  that doesn't include an `api::HashValue` in aptos-types and add a DB api to return it.

An API to get the BlockInfo by height should be straightforward since the sequence number on NewBlockEvent is the height.

We can return the transactions matching the block info directly from db as well.

@gregnazario

Reply:
Yeah, we could split it into two APIs if that is better.

I would have it returning the transactions matching the block info, except that then I would be reading every transaction until I found a block I wanted.

If there's an index to get the block for a block height, then this API wouldn't even be needed (except maybe for indexing)

Reply:
I'll make a get_block_info_by_height() later today